### PR TITLE
Fix erroneous closing of workflow tasks instead of heartbeating in some edge cases

### LINKED ISSUE #1262: Workflows stuck on local activity call
There are user reports pointing to workflows stuck on waiting for local activity results while activity is not in progress.

Workflow method stuck with stack dump:

```
"workflow-method-571893c0-821b-30e3-a437-80ce798dff42-1e8ec9fd-6426-4fd0-bc6d-8604beb6375d" #54994 prio=5 os_prio=0 cpu=1.70ms elapsed=559.87s tid=0x00007fa6a4004450 nid=0x58ac waiting on condition  [0x00007fa6e1ddf000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.15/Native Method)
	- parking to wait for  <0x0000000773341a78> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.15/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.15/AbstractQueuedSynchronizer.java:2081)
	at io.temporal.internal.sync.WorkflowThreadContext.yield(WorkflowThreadContext.java:89)
	at io.temporal.internal.sync.WorkflowThreadImpl.yield(WorkflowThreadImpl.java:412)
	at io.temporal.internal.sync.WorkflowThread.await(WorkflowThread.java:45)
	at io.temporal.internal.sync.CompletablePromiseImpl.getImpl(CompletablePromiseImpl.java:84)
	at io.temporal.internal.sync.CompletablePromiseImpl.get(CompletablePromiseImpl.java:74)
	at io.temporal.internal.sync.ActivityStubBase.execute(ActivityStubBase.java:44)
	at io.temporal.internal.sync.LocalActivityStubImpl.execute(LocalActivityStubImpl.java:29)
	at io.temporal.internal.sync.LocalActivityInvocationHandler.lambda$getActivityFunc$0(LocalActivityInvocationHandler.java:72)
	at io.temporal.internal.sync.LocalActivityInvocationHandler$$Lambda$1955/0x00000008012df440.apply(Unknown Source)
	at io.temporal.internal.sync.ActivityInvocationHandlerBase.invoke(ActivityInvocationHandlerBase.java:70)
	at com.sun.proxy.$Proxy128.publishPayload(Unknown Source)
```

History doesn't contain a marker for "publishPayload" local activity. History does contain workflow task completion with other local activity markers that were supposed to be executed before the activity in question.
Thread dump doesn't show "publishPayload" local activity currently in progress.

## Root cause analysis

Consider a scenario: 
Workflow task executes a sequence of local activities synchronously one by one. The whole chain of local activities doesn't fit into our Workflow Task Timeout and SDK has to perform a "Workflow Task Heartbeat". The last event loop that fits before the heartbeat scheduled a new local activity. Because of a `break` here: https://github.com/temporalio/sdk-java/blob/5079086e065adc64aa3d17e505482785f0f51dfe/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowRunTaskHandler.java#L284
this local activity command will never be taken by the state machines and accounted for in the outstanding local activities counter.

This leads to the counter of outstanding local activities of 0 in case of synchronous consecutive executions. Instead of performing a heartbeat, we close the workflow task, which leads to a stuck workflow.

## Reproduction

The problem is tough to reproduce. To make it happen, Workflow Task Heartbeat Timeout should happen DURING execution of workflow code (during eventLoop call) and this execution should schedule a local activity and this local activity should be the only outstanding local activity.

Reproduction uses Thread.sleep inside workflow code (should never be used in production code) to increase the chances of hitting the expiration of workflow task heartbeat during execution of workflow code.

## Context

Related user posts:
https://community.temporal.io/t/workflow-timedout-due-to-activity-blocked-on-feature-get/5136
https://community.temporal.io/t/activity-execution-is-stuck-on-blocked-on-feature-get/4247
https://community.temporal.io/t/activity-execution-is-blocked-on-blocked-on-feature-get/4391

### PR DESCRIPTION
**Pull request by Spikhalskiy**
## What was changed

Fixed situations when we didn't drain outstanding local activity requests which led to incorrect outstanding local activities counter value and erroneous closing of workflow tasks instead of performing a heartbeat.

Closes #1262
See #1262 for details

### CODE REVIEW

File: temporal-testing/src/main/java/io/temporal/internal/Issue.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package io.temporal.internal;
+
+/**
+ * Test annotated with {@link Issue} covers a specific problem or edge case and is usually crafted
+ * carefully to hit the right conditions to reproduce the problem. Please specify a link to a ticket
+ * that describes what this test intends to cover.
+ */
+public @interface Issue {
```
Comment (Sushisource) [inline temporal-testing/src/main/java/io/temporal/internal/Issue.java:28]: This is neat

File: temporal-sdk/src/test/java/io/temporal/workflow/activityTests/LocalActivityGettingScheduledRightBeforeWorkflowTaskHeartbeatTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.junit.Rule;
+import org.junit.Test;
+
+@Issue("https://github.com/temporalio/sdk-java/issues/1262")
+public class LocalActivityGettingScheduledRightBeforeWorkflowTaskHeartbeatTest {
+  private static final Duration WORKFLOW_TASK_TIMEOUT = Duration.ofSeconds(5);
+  private static final Duration SLEEP_DURATION =
+      Duration.ofMillis(800); // << 1000 to don't hit deadlock detection
```
Comment (Sushisource) [inline temporal-sdk/src/test/java/io/temporal/workflow/activityTests/LocalActivityGettingScheduledRightBeforeWorkflowTaskHeartbeatTest.java:46]:
```suggestion
      Duration.ofMillis(800); // << 1000 to avoid deadlock detection
```

**Review by Sushisource — APPROVED** [review]
